packages/bundlers/default/src/DefaultBundler.js

```diff
@@ -1119,6 +1122,23 @@ function createIdealGraph(
       }
     }
   }
+
+  // Fix asset order in source bundles as they are likely now incorrect after shared bundle deletion
+  if (modifiedSourceBundles.size > 0) {
+    let assetOrderMap = new Map(assets.map(a => [a, assets.indexOf(a)]));
+
+    for (let bundle of modifiedSourceBundles) {
+      bundle.assets = new Set(
+        [...bundle.assets].sort((a, b) => {
+          let aIndex = nullthrows(assetOrderMap.get(a));
+          let bIndex = nullthrows(assetOrderMap.get(b));
+
+          return aIndex - bIndex;
+        }),
+      );
+    }
+  }
```

Comment:
Wonder if you could use BitSet for `bundle.assets` here? Then when you do `bundle.assets.add` it would just be setting a bit corresponding to the position within `assets`. When we take the ideal graph and convert it back to the legacy graph we could iterate in asset order (as `BitSet#values()` does).

Reply:
Funny you mention this because I considered that approach as well but was worried this iteration cost might not be worth it. You've inspired me to try it though. Will report back.

Reply:
@devongovett Just realised it'd be hard to use the BitSet here as we need to create bundles before we know the full list of assets as we add them during the traverse.

Reply:
You could create the bitset earlier maybe

Reply:
After migrating all asset sets to BitSets I actually saw a drop in perf. This is due to the values() extraction being too expensive. I think there's potential to use BitSet more widely in the bundler but it would require a larger rethink to optimise around its benefits. 

tl;dr let's just sort for now.